src/charm.py

```diff
@@ -9,93 +9,377 @@
+        if not self.unit.is_leader:
+            return
+        # update values in the databag
+        self._update_active_relations(event, "active")
+
+    def _update_active_relations(self, event: RelationEvent, status: str):
+        """Update the relation status in the peer-relation databag."""
+        self.set_secret("app", f"relation-{ event.relation.id}", status)
+
+    def _on_peer_relation_changed(self, _: RelationEvent):
+        """Handle the peer relation changed event."""
+        if not self.unit.is_leader:
+            return
+        removed_relations = []
+        # check for relation that has been removed
+        for k, v in self.app_peer_data.items():
+            if "relation-" in k:
+                if v == "broken":
+                    removed_relations.append(k)
+
+        if removed_relations:
+            # update configuration
+            self._on_config_changed(None)
+            # update relation status to removed
+            for relation in removed_relations:
+                self.set_secret("app", relation, "removed")
+
+    @property
+    def database(self):
+        """Return the configured database name."""
+        return self.get_secret("app", "database")
+
+    @property
+    def topic(self):
+        """Return the configured topic name."""
+        return self.get_secret("app", "topic")
+
+    @property
+    def extra_user_roles(self):
+        """Return the configured user-extra-roles parameter."""
+        return self.get_secret("app", "extra-user-roles")
+
+    @property
+    def is_database_related(self):
+        """Return if a relation with database is present."""
+        possible_relations = [
+            self._check_for_credentials(self.mysql.relations),
+            self._check_for_credentials(self.postgresql.relations),
+            self._check_for_credentials(self.mongodb.relations),
+        ]
+        return any(possible_relations)
+
+    def _check_for_credentials(self, relations) -> bool:
+        """Check if credentials are present in the relation databag."""
+        for relation in relations:
+            if (
+                "username" in relation.data[relation.app]
+                and "password" in relation.data[relation.app]
+            ):
+                return True
+        return False
+
+    @property
+    def is_kafka_related(self):
+        """Return if a relation with kafka is present."""
+        return self._check_for_credentials(self.kafka.relations)
+
+    @property
+    def app_peer_data(self) -> Dict:
+        """Application peer relation data object."""
+        relation = self.model.get_relation(PEER)
+        if not relation:
+            return {}
+
+        return relation.data[self.app]
+
+    @property
+    def unit_peer_data(self) -> Dict:
+        """Peer relation data object."""
+        relation = self.model.get_relation(PEER)
+        if relation is None:
+            return {}
+
+        return relation.data[self.unit]
```

Comment:
**nitpick:** These aren't strictly-speaking `Dict` types. Change to `MutableMapping[str, str]`